roachtest: sqlsmith/setup=seed-vec/setting=vec failed

[(roachtest).sqlsmith/setup=seed-vec/setting=vec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1912287&tab=buildLog) on [master@61f18db7dd9a054d9a4648f67546202f760b5000](https://github.com/cockroachdb/cockroach/commits/61f18db7dd9a054d9a4648f67546202f760b5000):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed-vec/setting=vec/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: (("$4" IS DISTINCT FROM CAST(NULL AS DATE)) OR CAST(NULL AS DATE)) OR ("$8" AND "$8"): incompatible OR argument type: date
		stmt:
		SELECT
			tab_1269._interval AS col_4488,
			tab_1269._int8 AS col_4489,
			tab_1269._date AS col_4490,
			tab_1269._bool AS col_4491,
			tab_1269._uuid AS col_4492,
			tab_1269._float8 AS col_4493,
			tab_1269._decimal AS col_4494,
			tab_1269._string AS col_4495,
			tab_1269._int2 AS col_4496,
			tab_1269._bytes AS col_4497,
			tab_1269._int8 AS col_4498,
			tab_1269._timestamp AS col_4499,
			tab_1269._timestamptz AS col_4500
		FROM
			defaultdb.public.seed_vec@[0] AS tab_1269
		WHERE
			((tab_1269._date = tab_1269._date) OR (tab_1269._bool AND tab_1269._bool));

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed-vec/setting=vec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1912287&tab=artifacts#/sqlsmith/setup=seed-vec/setting=vec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed-vec%2Fsetting%3Dvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed-vec/setting=vec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1914617&tab=buildLog) on [master@a481ee62aba2ca146c3c90155f393beaa591ddda](https://github.com/cockroachdb/cockroach/commits/a481ee62aba2ca146c3c90155f393beaa591ddda):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed-vec/setting=vec/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: (("$0" IS DISTINCT FROM CAST(NULL AS FLOAT8)) OR CAST(NULL AS FLOAT8)) OR "$5": incompatible OR argument type: float
		stmt:
		SELECT
			tab_3264._interval AS col_11266,
			tab_3264._string AS col_11267,
			((tab_3264._float8 <= tab_3264._float8) OR tab_3264._bool) AS col_11268,
			tab_3266._date AS col_11269,
			tab_3264._decimal AS col_11270,
			tab_3264._timestamp AS col_11271,
			tab_3264._uuid AS col_11272,
			tab_3264._timestamptz AS col_11273,
			tab_3266._date AS col_11274
		FROM
			defaultdb.public.seed_vec@seed_vec__int8__float8__date_idx AS tab_3264
			JOIN defaultdb.public.seed_vec AS tab_3265
				JOIN defaultdb.public.seed_vec AS tab_3266 ON (tab_3265._int8) = (tab_3266._int8) ON
					(tab_3264._date) = (tab_3266._date)
					AND (tab_3264._date) = (tab_3265._date)
					AND (tab_3264._date) = (tab_3265._date)
		WHERE
			(tab_3264._bool AND tab_3264._bool)
		LIMIT
			95:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed-vec/setting=vec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1914617&tab=artifacts#/sqlsmith/setup=seed-vec/setting=vec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed-vec%2Fsetting%3Dvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed-vec/setting=vec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1916017&tab=buildLog) on [master@9410f923dc346ba3c87fa129c39c8fdf59e601d0](https://github.com/cockroachdb/cockroach/commits/9410f923dc346ba3c87fa129c39c8fdf59e601d0):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed-vec/setting=vec/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: "$8" AND (("$3" IS DISTINCT FROM CAST(NULL AS FLOAT8)) OR CAST(NULL AS FLOAT8)): incompatible OR argument type: float
		stmt:
		SELECT
			(tab_596._bool AND (tab_596._float8 <= tab_596._float8)) AS col_2197,
			tab_596._timestamptz AS col_2198,
			tab_596._int8 AS col_2199,
			tab_596._int4 AS col_2200,
			tab_596._date AS col_2201,
			tab_596._decimal AS col_2202,
			tab_596._interval AS col_2203,
			tab_596._float8 AS col_2204,
			tab_596._bytes AS col_2205,
			tab_596._timestamp AS col_2206
		FROM
			defaultdb.public.seed_vec@seed_vec__int8__float8__date_idx AS tab_596
		WHERE
			tab_596._bool
		ORDER BY
			tab_596._bool
		LIMIT
			7:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed-vec/setting=vec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1916017&tab=artifacts#/sqlsmith/setup=seed-vec/setting=vec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed-vec%2Fsetting%3Dvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed-vec/setting=vec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1918169&tab=buildLog) on [master@96f8f33dceeec0ad6168e2dffa98b952892bdf7b](https://github.com/cockroachdb/cockroach/commits/96f8f33dceeec0ad6168e2dffa98b952892bdf7b):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed-vec/setting=vec/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: "$8" OR (("$5" IS DISTINCT FROM CAST(NULL AS TIMESTAMP)) OR CAST(NULL AS TIMESTAMP)): incompatible OR argument type: timestamp
		stmt:
		SELECT
			tab_795._interval AS col_2741
		FROM
			defaultdb.public.seed_vec@[0] AS tab_795
		WHERE
			(tab_795._bool OR (tab_795._timestamp <= tab_795._timestamp))
		ORDER BY
			tab_795._uuid DESC
		LIMIT
			33:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed-vec/setting=vec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1918169&tab=artifacts#/sqlsmith/setup=seed-vec/setting=vec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed-vec%2Fsetting%3Dvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=seed-vec/setting=vec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1920699&tab=buildLog) on [master@425eaa8fb05fc32b2c42827b85338daa52f4177c](https://github.com/cockroachdb/cockroach/commits/425eaa8fb05fc32b2c42827b85338daa52f4177c):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed-vec/setting=vec/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: (("$11" IS NOT DISTINCT FROM CAST(NULL AS BYTES)) AND CAST(NULL AS BYTES)) AND "$8": incompatible AND argument type: bytes
		stmt:
		SELECT
			tab_198._interval AS col_730,
			tab_198._float8 AS col_731,
			tab_198._int4 AS col_732,
			tab_198._timestamp AS col_733,
			((tab_198._bytes < tab_198._bytes) AND tab_198._bool) AS col_734,
			tab_198._bytes AS col_735,
			tab_198._decimal AS col_736,
			tab_198._int8 AS col_737,
			tab_198._date AS col_738,
			tab_198._uuid AS col_739,
			tab_198._int8 AS col_740,
			tab_198._string AS col_741,
			tab_198._timestamptz AS col_742
		FROM
			defaultdb.public.seed_vec@seed_vec__int8__float8__date_idx AS tab_198
		LIMIT
			87:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed-vec/setting=vec](https://teamcity.cockroachdb.com/viewLog.html?buildId=1920699&tab=artifacts#/sqlsmith/setup=seed-vec/setting=vec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed-vec%2Fsetting%3Dvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/sqlsmith
+++ b/pkg/sql/logictest/testdata/logic_test/sqlsmith
@@ -141,3 +141,12 @@ NULL  1984-01-07 00:00:00 +0000 +0000  NULL  f96fd19a-d2a9-4d98-81dd-97e3fc2a45d
 NULL  1984-01-07 00:00:00 +0000 +0000  NULL  f96fd19a-d2a9-4d98-81dd-97e3fc2a45d2
 NULL  1984-01-07 00:00:00 +0000 +0000  NULL  f96fd19a-d2a9-4d98-81dd-97e3fc2a45d2
 NULL  1984-01-07 00:00:00 +0000 +0000  NULL  f96fd19a-d2a9-4d98-81dd-97e3fc2a45d2
+
+# Regression: #48267 (invalid opt transformation of OR with NULL)
+statement ok
+CREATE TABLE t (d) AS VALUES ('2001-01-01'::DATE)
+
+query T
+SELECT * FROM t WHERE (d = d) OR (d = d)
+----
+2001-01-01 00:00:00 +0000 +0000

--- a/pkg/sql/opt/norm/rules/scalar.opt
+++ b/pkg/sql/opt/norm/rules/scalar.opt
@@ -377,8 +377,8 @@
 )
 =>
 (Or
-    (IsNot $left (Null $typ:(TypeOf $left)))
-    (Null $typ)
+    (IsNot $left (Null (TypeOf $left)))
+    (Null (BoolType))
 )
 
 # SimplifySameVarInequalities converts `x != x` and other inequality
@@ -391,6 +391,6 @@
 )
 =>
 (And
-    (Is $left (Null $typ:(TypeOf $left)))
-    (Null $typ)
+    (Is $left (Null (TypeOf $left)))
+    (Null (BoolType))
 )

--- a/pkg/sql/opt/norm/testdata/rules/scalar
+++ b/pkg/sql/opt/norm/testdata/rules/scalar
@@ -1868,7 +1868,7 @@ project
  │    ├── columns: k:1!null
  │    └── key: (1)
  └── projections
-      └── (k:1 IS DISTINCT FROM CAST(NULL AS INT8)) OR CAST(NULL AS INT8) [as="?column?":7, outer=(1)]
+      └── (k:1 IS DISTINCT FROM CAST(NULL AS INT8)) OR CAST(NULL AS BOOL) [as="?column?":7, outer=(1)]
 
 # --------------------------------------------------
 # SimplifySameVarInequalities
@@ -1946,4 +1946,4 @@ project
  │    ├── columns: k:1!null
  │    └── key: (1)
  └── projections
-      └── (k:1 IS NOT DISTINCT FROM CAST(NULL AS INT8)) AND CAST(NULL AS INT8) [as="?column?":7, outer=(1)]
+      └── (k:1 IS NOT DISTINCT FROM CAST(NULL AS INT8)) AND CAST(NULL AS BOOL) [as="?column?":7, outer=(1)]

